modules/core/packer/roles/telegraf/files/run-telegraf.sh

```diff
@@ -0,0 +1,191 @@
+  local readonly name="$1"
+
+  if [[ ! $(command -v ${name}) ]]; then
+    log_error "The binary '${name}' is required by this script but is not installed or in the system's PATH."
+    exit 1
+  fi
+}
+
+function wait_for_consul {
+  local consul_leader
+
+  for (( i=1; i<="$MAX_RETRIES"; i++ )); do
+    consul_leader=$(
+      curl -sS http://localhost:8500/v1/status/leader 2> /dev/null || echo "failed"
+    )
+
+    if [[ "${consul_leader}" = "failed" ]]; then
+      log_warn "Failed to find Consul cluster leader. Will sleep for $SLEEP_BETWEEN_RETRIES_SEC seconds and try again."
+      sleep "$SLEEP_BETWEEN_RETRIES_SEC"
+    else
+      log_info "Found Consul leader at ${consul_leader}"
+      return
+    fi
+  done
+
+  log_error "Failed to detect Consul agent after $MAX_RETRIES retries. Did you start a Consul agent before running the script?"
+  exit 1
+}
+
+
+function consul_kv {
+  local readonly path="${1}"
+  local value
+  value=$(consul kv get "${path}") || exit $?
+  log_info "Consul KV Path ${path} = ${value}"
+  echo -n "${value}"
+}
+
+function consul_kv_with_default {
+  local readonly path="${1}"
+  local readonly default="${2}"
+  local value
+  value=$(consul kv get "${path}" || echo -n "${default}") || exit $?
+  log_info "Consul KV Path ${path} = ${value}"
+  echo -n "${value}"
+}
+
+function main {
+  local type=""
+  local consul_prefix="terraform/"
+  local skip_template="false"
+  local conf_template="/etc/telegraf/telegraf.conf.template"
+  local conf_out="/etc/telegraf/telegraf.conf"
+
+  local readonly service_override_dir="/etc/systemd/system/telegraf.service.d"
+
+  while [[ $# > 0 ]]; do
+    local key="$1"
+
+    case "$key" in
+      --type)
+        assert_not_empty "$key" "$2"
+        type="$2"
+        shift
+        ;;
+      --consul-prefix)
+        assert_not_empty "$key" "$2"
+        consul_prefix="$2"
+        shift
+        ;;
+      --skip-template)
+        skip_template="false"
+        ;;
+      --conf-template)
+        assert_not_empty "$key" "$2"
+        conf_template="$2"
+        shift
+        ;;
+      --conf-out)
+        assert_not_empty "$key" "$2"
+        conf_out="$2"
+        shift
+        ;;
+      --help)
+        print_usage
+        exit
+        ;;
+      *)
+        log_error "Unrecognized argument: $key"
+        print_usage
+        exit 1
+        ;;
+    esac
+
+    shift
+  done
+
+  if [[ "$type" != "vault" && "$type" != "consul" && "$type" != "nomad_server" && "$type" != "nomad_client" ]]; then
+    log_error "Invalid type set."
+    exit 1
+  fi
+
+  assert_is_installed "consul"
+  assert_is_installed "consul-template"
+
+  wait_for_consul
+
+  local readonly enabled=$(consul_kv_with_default "${consul_prefix}telegraf/enabled" "no")
+
+  if [[ "$enabled" != "yes" ]]; then
+    log_info "Telegraf is not enabled"
+  else
+    if [[ "$skip_template" == "false" && -f "$conf_template" ]]; then
+      consul-template -template "$conf_template:$conf_out" -once
```

Comment:
Might need to point `consul-template` to the configuration directory for `consul-template`.

Reply:
Hmm are you referring to something like this: `consul-template -config /opt/consul-template/config/ -template "/etc/telegraf/telegraf.conf.template:/etc/telegraf/telegraf.conf" -once`?

The above works without setting `-config` at all actually.

On the other hand, setting it like the above fails:
```
2018/06/11 03:40:02.870838 [WARN] (view) vault.read(auth/token/lookup-self): vault.read(auth/token/lookup-self): Error making API request.

URL: GET https://vault.service.consul:8200/v1/auth/token/lookup-self
Code: 400. Errors:

* missing client token (retry attempt 1 after "250ms")
```

Performing a manual `vault login` will make the command with `-config` work again, but the end result  seems to be the same.

Reply:
Strange.

The run consul template script will be the one getting the initial token.

Reply:
To be fair, i didn't run the `-config` in the user data, but rather run manually on SSH login after the user data has already been done.

Let me try putting this into user data and start up one `consul` with this change.

Reply:
Hmm okay it's similar, logs from user-data:
```
2018/06/11 04:09:15.980480 [INFO] (runner) rendered "/etc/telegraf/telegraf.conf.template" => "/etc/telegraf/telegraf.conf"
2018/06/11 04:09:16.087859 [WARN] (view) vault.read(auth/token/lookup-self): vault.read(auth/token/lookup-self): Error making API request.

URL: GET https://vault.service.consul:8200/v1/auth/token/lookup-self
Code: 400. Errors:

* missing client token (retry attempt 1 after "250ms")
```

It was able to generate the config file from template, but subsequently returns non-zero exit code, causing the rest of the steps to fail.

Do you want to investigate this further, or you okay to accept without setting the `-config` for now?

Reply:
Let's leave it for now. We can fix it later if anything breaks.